R/submit-model.R

```diff
@@ -356,6 +356,38 @@ check_mode_argument <- function(.mode) {
     stop(BBI_EXE_MODE_INVALID_ERR_MSG, call. = FALSE)
   }
 
+  if (isTRUE(getOption("bbr.DEV_skip_system_mode_checks"))) {
+    # ^ Allow dry-run tests to skip the system mode checks.
+    return(invisible(TRUE))
+  }
+
+  # All of the remaining checks depend on system details.
+
+  if (identical(.mode, "slurm") && !test_bbi_version(.min_version = "3.4.0")) {
+    stop(
+      "Installed bbi version is ", bbi_version(),
+      ", but .mode='slurm' requires at least version 3.4.0"
+    )
+  }
+
+  if (identical(.mode, "sge")) {
+    qsub <- unname(Sys.which("qsub"))
+    if (identical(qsub, "")) {
+      stop(".mode='sge' but qsub is not available on system")
+    }
+
+    # This guard is Metworx (or really ParallelCluster) specific. Slurm ships
+    # with a qsub shim. That leads to a confusing situation where 'bbi nonmem
+    # run sge ...' will not abort upfront, but the execution doesn't entirely
+    # match the expected behavior (especially under --parallel).
+    if (identical(qsub, "/opt/slurm/bin/qsub")) {
```

Comment:
I like this idea, but I want to think through situations where a user may have been previously using bbr/bbi with Slurm via the `qsub` shim... If they now upgrade to this (upcoming) bbr release, will it be as simple as this?

* They start seeing this error
* They need to upgrade `bbi` and switch their code to explicitly call `.mode = "slurm` (or set the option)
* It should work as it did before

It seems like there may be some nuance that I'm missing though. Maybe we don't need to try to go too far to cover these somewhat hacky cases, but I want to at least discuss it.

Reply:
> [...] I want to think through situations where a user may have been previously using bbr/bbi with Slurm via the `qsub` shim...

I agree this is worth thinking through.  Here's the rationale from ba073b1b for this (admittedly aggressive) approach:

```
...

Slurm includes a qsub shim that has some behavioral differences when
used under 'bbi nonmem run sge ...' (e.g., output files going to the
parent directory).  That shim would very likely cause confusion when
people on Slurm machines forget to override the "sge" default.

Prevent that confusion on Metworx or other systems using
ParallelCluster by aborting if "sge" is specified for .mode and qsub
is Slurm's shim.  That's aggressive and perhaps will be triggered by
people that have already been making of the shim.  However, it's
better to be strict about it because it's not an option we intend to
support.
```

To rephrase and expand a bit:

 1) I'm concerned about users unintentionally selecting "sge" given the default

 2) I was seeing some pretty wonky results during my (unsystematic) testing  of `bbi nonmme run sge` with Slurm's qsub shim.  Two things that come to mind (both with `--parallel`, I think) is 1) some generated files being placed outside the output directory and 2) the `--threads` value not being respected.

> If they now upgrade to this (upcoming) bbr release, will it be as simple as this? [...]

I think that captures it, though "should work as it did before" will hopefully not include any weirdness like what I mentioned above.

> Maybe we don't need to try to go too far to cover these somewhat hacky cases, but I want to at least discuss it.

Definitely worth discussing, and I'm open to other ideas here.  As mentioned above, I would be concerned with removing the guard entirely.

If some users rely on this shim heavily, they could abuse the `bbr.DEV_skip_system_mode_checks` option to prevent the guard from firing, at least in the short term.  If turns out that the shim is relied on more than we think, we can elevate that to a non-dev option and document it.